text/0000-seth-rpc-extensions.md

```diff
@@ -0,0 +1,347 @@
+- Start Date: 2018-07-30
+- RFC PR:
+- Sawtooth Issue:
+
+# Summary
+[summary]: #summary
+
+The Seth JSON-RPC API is an implementation of the Ethereum JSON-RPC API, used
+for cross-compatibility with existing Ethereum clients. The Ethereum JSON-RPC
+API lacks several areas of functionality that require the existing Seth client
+to talk directly to the Sawtooth REST API. This RFC proposes several JSON-RPC
+API extensions that will allow the CLI to become a thin client that builds on
+top of the `web3` library.
+
+# Motivation
+[motivation]: #motivation
+
+Effecting this RFC proposal will bring us in line with existing extensions to
+the Ethereum JSON-RPC API, allowing current Ethereum clients to work with Seth
+without modification.
+
+Moving the CLI to becoming a thin client will allow a simplified architecture
+that stores private keys in one place, and will drastically reduce code
+complexity. It will also eliminate the need for clients to understand how to
+talk to the Sawtooth REST API.
+
+An additional benefit will be that dogfooding our JSON-RPC API will ensure that
+the implementation is tested and up-to-date with other implementations.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Managing Accounts
+
+The [parity] project has faced the same issue of allowing account management
+via the [JSON-RPC API][personal-api] that it represents. This RFC proposes
+implementing a compatible API that presents these endpoints:
+
+ - personal_listAccounts
+ - personal_newAccount
+ - personal_unlockAccount
+ 
+Additionally, in order to facilitate import existing accounts, the API will
+present this method that [geth][import-raw-key] has implemented:
+
+ - personal_importRawKey
+
+[parity]: https://www.parity.io/
+[import-raw-key]: https://github.com/ethereum/go-ethereum/wiki/Management-APIs#personal_importrawkey
+[personal-api]: https://wiki.parity.io/JSONRPC-personal-module.html
+
+## Managing Permissions
+
+Two new endpoints will be added for general permissions management, and one
+existing endpoint will be extended to handle permissions. The new endpoints
+will be `seth_getPermissions` and `seth_setPermissions` and will allow general
+retrieval and setting of permissions.
+
+The existing `eth_sendTransaction` endpoint will be extended to allow a
+`permissions` argument.
+
+Finally, the proposed `personal_importRawKey` and `personal_newAccount`
+endpoints will accept a `permissions` argument as well.
+
+## Contract Chaining
+
+Sawtooth has the concept of a limited set of input and output address that a
+transaction is limited to reading from/writing to. That concept doesn't exist
+in Ethereum, and makes the concept of contract chaining (e.g. creating a
+contract from another contract) difficult. The `eth_sendTransaction` endpoint
+will be extended to allow an argument that sets the list of inputs and outputs
+to be the entire Sawtooth Seth address space, enabling general contract
+chaining.
```

Comment:
Can you expand on this? How do inputs/outputs default today? What implications are there in enabling the entire address space ... performance? ... security? How do the transaction input/output addresses interact or not with permissions?

Reply:
After talking with @aludvik, I think we can default to allowing contract chaining (i.e., transactions are simply limited to the seth address namespace). This is what Ethereum-based tools expect, and we can then tackle restricting the address space for contract chaining as a separate ticket. So, this section has been removed.